INTRODUCTION.md

```diff
@@ -426,7 +426,8 @@ There are three types of guarantees that the idempotent producer can satisfy:
               This guarantee is enabled by default, but may be disabled
               by setting `enable.gapless.guarantee` if individual message
               failure is not a concern.
-
+              Message timeouts (`message.timeout.ms`) will cause gaps which
+              are permitted by the gap-less guarantee.
```

Comment:
I find this section confusing to grok, and suggest framing it 
a bit differently as below (you may want to improve on my 
choice of words, which is not perfect..).

```
The idempotent producer provides two guarantees:

 * Exactly-once - a message is only written to the log once.
 * Ordering - a series of messages are written to the log in the
              order they were produced.

Additionally, the behavior in the event of message timeouts 
(`message.timeout.ms`) is configurable via the 
`enable.gapless.guarantee` configuration parameter. By 
default, this is set to `true`, which guarantees that a 
series of messages are written once and in order
without risk of skipping messages. The sequence of
message may be cut short and fail before all messages
are written, but may not fail individual messages in the
series. It's important to note that with the gapless 
guarantee enabled, a message delivery failure will result
in a fatal producer error. If individual message failure is
not a concern, `enable.gapless.guarantee` can be set
to `false`.

Note: The idempotent producer feature only applies to
the producer - it does NOT cover the exactly once 
consume case.
```

Reply:
I don't understand the reason for this config?

Reply:
@ijuma There are certain error cases (any non-idempotent permanent error) which may cause messages to permanently fail producing, which will lead to "gaps" in the message series.
E.g.: messages 1,2,3,4,5 are produced by the application, 1,2 are produced, 3,4 fail due to invalid message (size, or whatever), 5 is produced, leading to 1,2,5 being persisted in the log.
With the gap-less guarantee we provide the application with an additional guarantee (over exactly once and ordering) to stop producing if it would result in gaps in the message series, and instead raise a fatal error.

Message timeouts are not covered by the gap-less guarantee.

Reply:
Interesting. That could be useful. The naming is a bit unfortunate though because of the timeouts edge case.

Reply:
I believe it is useful until we add transactions, which is a better and more resilient way to solve this.
We could leave it out, or mark it as experimental, and make the default gapless.guarantee=false (which is the Java producer behaviour), until we know what to do with this.

Reply:
@mhowlett 

Thanks for the improved verbiage.

This bit is not true though, since message timeouts are okay even with gapless.guarantee enabled:

> Additionally, the behavior in the event of message timeouts 
> (`message.timeout.ms`) is configurable via the 
> `enable.gapless.guarantee`

Reply:
If this config hasn't been in any release and we are not sure about it, I'd vote for not including it in 1.0.

Reply:
For reference, this is the configuration property doc string:
```
        { _RK_GLOBAL|_RK_PRODUCER|_RK_HIGH, "enable.gapless.guarantee",
          _RK_C_BOOL,
          _RK(eos.gapless),
          "When set to `true`, any error that could result in a gap "
          "in the produced message series when a batch of messages fails, "
          "will raise a fatal error (ERR__GAPLESS_GUARANTEE) and stop "
          "the producer. "
          "Requires `enable.idempotence=true`.",
          0, 1, 1 },
```